### ISSUE TITLE
Passing state between server and client?

### ISSUE THREAD

**Issue by nainy** [AlexKvazosMx/React-Starter#1]
I really like the approach this starter kit has to server rendering.

However, I was wondering if it would be possible to implement some sort of state sharing mechanism between the server and the client? 

Like here, for example: https://github.com/slmgc/react-isomorphic-immutable, although that approach seems hacky to me.

Any ideas? Thanks.

**Comment by AlexKvazos** [comment]
Hello @nainy,

This kit does not render the actual application on the server side, it just renders the essential <head> & <body> tags with the links to the css and js bundles. The rest of the applications is rendered on the clientside. If you look closely at the RenderController, we are not rendering the router with the matched component, we are just rendering a Layout component.

**Comment by nainy** [comment]
Yes, I understand that, and I like it.

But I think it would've been nice if one could fill the state tree on the server at least partially and then pass it to the client. For instance, I like to manipulate authentication and cookies on the server side, as I find it to be slightly more secure.

**Comment by AlexKvazos** [comment]
Okay, I do have this on one of my apps built from this kit, and here is what I did.

On the server, before the request reaches the controller, I populate the `req.user` object with a middleware. I require the state on the server and I put the user into it.

Once in the render controller, the state gets serialized into JSON and sent to the <Layout /> component. The state is stored in a  `window.__state__` object.

In the client, baobab creates the state from this json object, and the rest works like always.

**Comment by nainy** [comment]
Interesting.

Do you just fill the already created tree in the Container component in some life cycle method like `componentWillMount()` using context? I thought you cannot really pass a tree to a rooted higher-order component.

**Comment by AlexKvazos** [comment]
No, components do not come into play on the server. I simple feed data into the state, and on the RenderController I serialize it and embed it to the Layout component using `window.__state__`.  The client will unserialize this json and create a state from it.

**Comment by nainy** [comment]
Can you please show the code that does this?
